File: K Nearest Neighbors/Algorithms in Rust/knearest.rs
use std::collections::HashMap;
use std::cmp::Ordering;
use ndarray::{Array1, Array2};
use ndarray::s;
use ndarray_stats::DeviationExt;

// Define a struct to represent a data point
#[derive(Debug)]
struct DataPoint {
    features: Array1<f64>,
    label: String,
}

// Implement a function to calculate the Euclidean distance between two data points
fn euclidean_distance(a: &Array1<f64>, b: &Array1<f64>) -> f64 {
    let squared_diff: Array1<f64> = a - b;
    squared_diff.dot(&squared_diff).sqrt()
}

// Implement the KNN algorithm
fn knn(training_data: &[DataPoint], input: &Array1<f64>, k: usize) -> String {
    // Calculate the distances between the input and all training data points
    let distances: Vec<(f64, &DataPoint)> = training_data
        .iter()
        .map(|data| (euclidean_distance(&data.features, input), data))
        .collect();

    // Sort the distances in ascending order
    let sorted_distances = distances.into_iter().sorted_by(|a, b| a.0.partial_cmp(&b.0).unwrap());

    // Count the labels of the k nearest neighbors
    let mut label_counts: HashMap<&str, usize> = HashMap::new();
    for (_, data) in sorted_distances.take(k) {
        *label_counts.entry(&data.label).or_insert(0) += 1;
    }

    // Find the label with the highest count
    let (label, _) = label_counts
        .into_iter()
        .max_by(|(_, count1), (_, count2)| count1.cmp(count2))
        .unwrap();

    label.to_string()
}

fn main() {
    // Create some sample training data
    let training_data = vec![
        DataPoint {
            features: Array1::from(vec![2.0, 4.0]),
            label: String::from("A"),
        },
        DataPoint {
            features: Array1::from(vec![4.0, 6.0]),
            label: String::from("A"),
        },
        DataPoint {
            features: Array1::from(vec![1.0, 3.0]),
            label: String::from("B"),
        },
        DataPoint {
            features: Array1::from(vec![5.0, 7.0]),
            label: String::from("B"),
        },
    ];

    // Create an input data point
    let input = Array1::from(vec![3.0, 5.0]);

    // Perform KNN classification with k=3
    let predicted_label = knn(&training_data, &input, 3);

    println!("Input: {:?}", input);
    println!("Predicted Label: {}", predicted_label);
}

File: K Nearest Neighbors/Algorithms in Go/knearest.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type DataPoint struct {
	X float64
	Y float64
}

type Distance struct {
	Point  DataPoint
	Dist   float64
	Target string
}

type ByDistance []Distance

func (a ByDistance) Len() int           { return len(a) }
func (a ByDistance) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByDistance) Less(i, j int) bool { return a[i].Dist < a[j].Dist }

func main() {
	// Training data
	trainingData := []DataPoint{
		{2.0, 4.0},
		{4.0, 6.0},
		{2.5, 5.0},
		{2.2, 4.3},
		{3.5, 5.5},
		{1.8, 3.8},
		{4.5, 6.7},
		{3.3, 4.9},
	}

	// Test point
	testPoint := DataPoint{3.0, 5.0}

	// Number of nearest neighbors
	k := 3

	// Compute distances
	distances := make([]Distance, len(trainingData))
	for i, point := range trainingData {
		distances[i] = Distance{
			Point:  point,
			Dist:   euclideanDistance(point, testPoint),
			Target: "unknown",
		}
	}

	// Sort distances
	sort.Sort(ByDistance(distances))

	// Get k nearest neighbors
	neighbors := distances[:k]

	// Count targets
	targets := make(map[string]int)
	for _, neighbor := range neighbors {
		targets[neighbor.Target]++
	}

	// Get most common target
	var maxCount int
	var target string
	for t, count := range targets {
		if count > maxCount {
			maxCount = count
			target = t
		}
	}

	fmt.Println("Test point:", testPoint)
	fmt.Println("Predicted target:", target)
}

// Euclidean distance between two data points
func euclideanDistance(a, b DataPoint) float64 {
	dx := a.X - b.X
	dy := a.Y - b.Y
	return math.Sqrt(dx*dx + dy*dy)
}
